design/proposals/ks-registry-update-delete.md

```diff
@@ -0,0 +1,121 @@
+that will allow users to update and delete the local cache of their registry.
+This is intended to provide a clear path for users to obtain new libraries that
+have been added to the source registry, as well as new versions of old libraries
+in a registry.
+
+## Motivation
+
+The `ks` client tool currently allows user to manage _registries_—collections of
+_packages_ that can be installed and used in the local ksonnet application.
+
+Managing the local cache of a registry when the two diverge is a major painpoint
+in the `ks` client tool. In particular, `ks` currently does not have the ability
+either to update the registry, or to delete a registry, so when a new package is
+added to a remote registry, in order to obtain it, they must:
+
+1. figure out how registries work (_i.e._, via the registry specification in a
+   YAML file), and
+2. manually update their registry spec files (and possibly script it, as [this
+   user has done])
+
+Since users are now creating their own custom registries, we can't afford for
+this part of the workflow to be painful.
+
+The related issue is captured in [#237][237].
+
+## Goal
+
+* Make it trivial to additively refresh registries (_i.e._, most updates should
+  be one command).
+* Make it trivial to delete a registry (_i.e._, if we don't depend on anything
+  from a registry, it should take one command).
+* The `registry update` and `registry delete` commands should output a
+  trivially-consumable summary of changes that are being incorporated. (_e.g._,
+  "adding these packages", "removing these", "adding releases for these
+  packages")
+* Should alert the user if a `registry update` or `registry delete` is about to
+  make the application un-buildable. For example, if `registry delete`'ing the
+  registry would delete a version of a dependency you actually use, we should
+  warn the user before proceeding (perhaps requiring a `--force` flag).
+
+## Non Goals
+
+* Any sort of package refresh. Something like `ks pkg upgrade` is orthogonal to
+  the current expressed pain of users, and is complex enough to greatly slow the
+  speed of delivering the cause of this pain (namely, `ks registry update`).
+* Transitive closure resolution. Our package structure is very simple, so we
+  don't need to traverse dependencies of dependencies (yet!).
+
+## Proposal
+
+We propose to implement the following commands:
+
+```
+ks registry update <registry-name> [flags]
+
+Attempt to update our local cache of a package registry. This might include
+adding or removing packages from the cached registry, as well as adding or
+removing specific versions of a package. These packages can be installed into
+the current application with `ks pkg`.
+
+This command additionally provides sanity checks, and will fail in cases where
+updating would make the application un-buildable. For example, if 'registry
+update' would remove a package (or a version of a package) that the application
```

Comment:
To make registries easily upgradable, it appears registries should be structured in a way where its packages are version-able. While this may seem obvious, it's less clear when looking at the example registry [incubator](https://github.com/ksonnet/parts/tree/master/incubator) where there's no evidence of versioning for `mysql`, `redis`, etc.

Maybe not as a part of this specific proposal, but as a related goal we should think about how to structure packages in a way that is kind to versioning.  If this functionality already exists, there probably needs to be some documentation so it is clear to registry creators.

Reply:
Actually packages are versioned! See [here][version]. The version for git registries is a git refspec, in this case `master`. This was hastily done, and we should definitely think more about it. I'll file an issue to follow up.

[version]: https://github.com/ksonnet/parts/blob/master/incubator/registry.yaml#L5

Reply:
@jessicayuen actually, now that I think about it, I'm ok with our versioning semantics. I propose we actually should wrap the discussion about the versioning semantics into #232, which is when we'd generalize the current registry spec anyway.